### PR TITLE
Add constraints for aarch64: numpy==1.19.0 for Python<=3.8

### PR DESCRIPTION
**Pull request by terencehonles**
Downloads for:
- https://pypi.org/project/numpy/1.18.5/#files (no aarch64)
- https://pypi.org/project/numpy/1.19.0/#files (aarch64)

xref:
- https://github.com/scikit-learn/scikit-learn/pull/18782#issuecomment-723464240
- https://github.com/scikit-learn/scikit-learn/pull/18761 see: https://github.com/scikit-learn/scikit-learn/pull/18761/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -18,14 +18,29 @@ version = 0.6
 [options]
 python_requires = >=3.5
 install_requires =
-    numpy==1.13.3; python_version=='3.5' and platform_system!='AIX'
-    numpy==1.13.3; python_version=='3.6' and platform_system!='AIX'
-    numpy==1.14.5; python_version=='3.7' and platform_system!='AIX'
+    # NOTE: platform_system 'AIX' and platform_machine 'aarch64' are mutually
+    #       exclusive so they are specified by themself when setting their
+    #       numpy version.
```
Comment (terencehonles) [inline setup.cfg:23]: I'm fairly certain this is correct, but if someone knows better I'd definitely want their advice.

Comment (rgommers) [reply setup.cfg:23]: Yes that's correct.

File: setup.cfg [hunk 2]
```diff
@@ -18,14 +18,29 @@ version = 0.6
...
+
+    # platform_system=='AIX' requires numpy 1.16.0 (needed for Python < 3.8)
     numpy==1.16.0; python_version=='3.5' and platform_system=='AIX'
     numpy==1.16.0; python_version=='3.6' and platform_system=='AIX'
     numpy==1.16.0; python_version=='3.7' and platform_system=='AIX'
-    numpy==1.17.3; python_version=='3.8'
+
+    # platform_machine=='aarch64' requires numpy 1.19.0 (needed for Python < 3.9)
+    numpy==1.19.0; python_version=='3.5' and platform_machine=='aarch64'
```
Comment (rgommers) [inline setup.cfg:31]: This is incorrect, numpy 1.19 doesn't support Python 3.5 anymore. Can you change this to `1.18.5` (no wheel, but it does work)?

File: setup.cfg [hunk 3]
```diff
@@ -18,14 +18,29 @@ version = 0.6
...
+    #       exclusive so they are specified by themself when setting their
+    #       numpy version.
+
+    # platform_system=='AIX' requires numpy 1.16.0 (needed for Python < 3.8)
     numpy==1.16.0; python_version=='3.5' and platform_system=='AIX'
     numpy==1.16.0; python_version=='3.6' and platform_system=='AIX'
     numpy==1.16.0; python_version=='3.7' and platform_system=='AIX'
-    numpy==1.17.3; python_version=='3.8'
+
+    # platform_machine=='aarch64' requires numpy 1.19.0 (needed for Python < 3.9)
```
Comment (rgommers) [inline setup.cfg:30]: The first part isn't really true, and the "needed for Python < 3.9" is unnecessary and will go out of date. Can you change to:
```
# numpy 1.19 was the first minor release to provide aarch64 wheels
```

File: setup.cfg [hunk 4]
```diff
@@ -18,14 +18,29 @@ version = 0.6
...
+    numpy==1.19.0; python_version=='3.5' and platform_machine=='aarch64'
+    numpy==1.19.0; python_version=='3.6' and platform_machine=='aarch64'
+    numpy==1.19.0; python_version=='3.7' and platform_machine=='aarch64'
+    numpy==1.19.0; python_version=='3.8' and platform_machine=='aarch64'
```
Comment (rgommers) [inline setup.cfg:34]: For python 3.6-3.8 we want `1.19.2` as the minimum I think, that fixed quite a nasty bug (see https://github.com/numpy/numpy/releases/tag/v1.19.2).

File: setup.cfg (hunk 3 above)
Comment (rgommers) [reply setup.cfg:30]: And same for the AIX comment above, can you change that to
```
# AIX requires fixes contained in numpy 1.16 
```

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @terencehonles, this looks good to me - just needs a few small tweaks.

File: setup.cfg (hunk 2 above)
Comment (terencehonles) [reply setup.cfg:31]: I didn't think about them dropping Python 3.5. Should it just not be declared?

I know this library doesn't want to pull dependencies that previously existed so it doesn't yank support from downstream dependencies, but what is the proper stance in this case? Should it say that aarch64 requires Python 3.6? I can change it to `1.18.5` as suggested, but I don't know if it works either. I just know wheels were built in 1.19.

File: setup.cfg (hunk 3 above)
Comment (terencehonles) [reply setup.cfg:30]: Adjusting, but I was mainly trying to call out why it stops at a specific Python version and to help future contributors think about that when suggesting changes to this file.

File: setup.cfg (hunk 4 above)
Comment (terencehonles) [reply setup.cfg]: I altered your suggested comment to include this

File: setup.cfg (hunk 2 above)
Comment (terencehonles) [reply setup.cfg]: Poking around a little, and I don't think I have much time to really get to the bottom of it, but it looks like aarch could have been around for awhile: https://github.com/numpy/numpy/pull/11016

More recent request to build wheels:
- https://github.com/numpy/numpy/pull/15279
- https://github.com/numpy/numpy/issues/14886

I'm not sure if the goal of this library is to provide the versions needed to use wheels or if versions building from source are OK. If it's building from source `1.19.2` isn't required for Python 3.6-3.8, but if it's to build from a wheel `1.19.2` is required (just applying this same comment to the other versions).

Comment (rgommers) [reply setup.cfg]: I'm fairly sure `aarch64` in 1.18.5 worked, so I'd add that for completeness. In practice it won't matter much, py35 is ancient history.

Comment (terencehonles) [reply setup.cfg]: Ok, I've updated the config. Do you know if it should be earlier than 1.18.5? I'm not sure if that's the oldest version or if doesn't matter as much if it's built from source.

Comment (rgommers) [reply setup.cfg]: I think 1.18.5 is right. There were definitely some useful fixes in the 1.18.x branch, and requiring the most recent bugfix release for a minor version seems like the right thing to do.

**Comment by terencehonles** [comment]
@rgommers was there anything else you'd want changed? I believe we've addressed everything.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @terencehonles

**Comment by terencehonles** [comment]
> LGTM now, thanks @terencehonles

@rgommers Great! Thanks! What was the release schedule for this package? I didn't see it in the README so I figured I'd ask. I'm specifically asking because in scikit-learn/scikit-learn#18900 I'm proposing to use the changes I added in this PR.

**Comment by astrofrog** [comment]
I could do a release tomorrow

**Comment by rgommers** [comment]
That sounds great, thanks @astrofrog.

**Comment by astrofrog** [comment]
I've now released v0.7!

**Pull request merged**